src/TrackGenerator.cpp

```diff
@@ -821,44 +873,106 @@ void TrackGenerator::initializeTrackUIDs() {
+ *          that can be looped over in parallel without data races. This method
+ *          creates a 1D array of Track pointers where the tracks are arranged
+ *          by parallel group. If the geometry has a periodic BC, 6 periodic
+ *          groups are created; otherwise, 2 periodic groups are created. The
+ *          Track UIDs are also set to their index in the tracks by periodic
+ *          group array.
+ */
+void TrackGenerator::initializeTrackUIDs() {
 
+  Track* track;
+  bool periodic;
   int uid = 0;
-  int index;
-  _num_tracks_by_halfspace[0] = 0;
-  
-  /* Recalibrate the tracks to the origin and set the uid. Note that the 
-   * loop structure is unconventional in order to preserve a monotonically
-   * increasing track uid value in the Solver's tracks array. The tracks array
-   * is oriented such the tracks can be broken up into 4 sub arrays that are
-   * guaranteed to contain tracks that do not transport into other tracks both
-   * reflectively and periodically. This is done to guarantee reproducability
-   * in parallel runs. */
-  for (int azim_halfspace=0; azim_halfspace < 2; azim_halfspace++) {
-    for (int period_halfspace=0; period_halfspace < 3; period_halfspace++) {
-      for (int a=azim_halfspace*_num_azim/2;
-           a < (azim_halfspace+1)*_num_azim/2; a++) {
-        for (int i=0; i < _num_tracks[a]; i++) {
-
-          track = &_tracks[a][i];
-          index = track->getPeriodicTrackIndex();
-
-          /* Check if track UID should be set */
-          if (period_halfspace == 0 && index == 0) {
-            track->setUid(uid);
-            uid++;
-          }
-          else if (period_halfspace == 1 && index % 2 == 1) {
-            track->setUid(uid);
-            uid++;
-          }
-          else if (period_halfspace == 2 && index % 2 == 0 && index != 0) {
-            track->setUid(uid);
-            uid++;
-          }
+  int azim_group_id, periodic_group_id;
+  int track_azim_group_id, track_periodic_group_id;
+  int track_periodic_index;
+
+  /* If periodic boundary conditions are present, set the number of parallel
+   * track groups to 6; else, set the number of parallel track groups to 2. Also
+   * set the periodic boolean indicating whether periodic bcs are present. */
+  if (_geometry->getMinXBoundaryType() == PERIODIC ||
+      _geometry->getMinYBoundaryType() == PERIODIC) {
+    _num_parallel_track_groups = 6;
+    periodic = true;
+  }
+  else {
+    _num_parallel_track_groups = 2;
+    periodic = false;
+  }
+
+  /* Allocate memory for the num tracks by parallel group array */
+  _num_tracks_by_parallel_group = new int[_num_parallel_track_groups + 1];
+  _num_tracks_by_parallel_group[0] = 0;
+
+  /* Allocate memory for the tracks by parallel group array */
+  int num_tracks = getNumTracks();
+  _tracks_by_parallel_group = new Track*[num_tracks];
+
+  /* Loop over the parallel track groups */
+  for (int g = 0; g < _num_parallel_track_groups; g++) {
```

Comment:
This is fine for now, but eventually we will want to change this to be parallel over all tracks, using point-jacobi updating of angular fluxes